pkg/ottl/ottlfuncs/func_slice_to_map.go

```diff
@@ -32,80 +32,107 @@ func sliceToMapFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ot
+				return pcommon.Map{}, fmt.Errorf("element %d: could not extract key from element: %w", i, err)
 			}
-
-			k, ok := extractedKey.(string)
-			if !ok {
-				return nil, errors.New("extracted key attribute is not of type string")
+			if extractedKey.Type() != pcommon.ValueTypeStr {
+				return pcommon.Map{}, fmt.Errorf("element %d: extracted key attribute is not of type string", i)
 			}
-			key = k
+			key = extractedKey.Str()
 		}
 
-		if valuePath.IsEmpty() {
-			if err := m.PutEmpty(key).FromRaw(e); err != nil {
-				return nil, fmt.Errorf("could not convert value from element: %w", err)
+		if useValuePath {
+			extractedValue, err := extractPcommonValue(elem.Map(), valuePath.Get())
+			if err != nil {
+				return pcommon.Map{}, fmt.Errorf("could not extract value from element: %w", err)
 			}
-			continue
+			value = extractedValue
 		}
-		extractedValue, err := extractValue(e, valuePath.Get())
+
+		err := putValue(value, m, key)
 		if err != nil {
-			return nil, fmt.Errorf("could not extract value from element: %w", err)
-		}
-		if err = m.PutEmpty(key).FromRaw(extractedValue); err != nil {
-			return nil, fmt.Errorf("could not convert value from element: %w", err)
+			return pcommon.Map{}, err
 		}
 	}
 	return m, nil
 }
 
-func extractValue(v map[string]any, path []string) (any, error) {
+func putValue(value pcommon.Value, m pcommon.Map, key string) error {
+	switch value.Type() {
+	case pcommon.ValueTypeStr:
+		m.PutStr(key, value.Str())
+	case pcommon.ValueTypeInt:
+		m.PutInt(key, value.Int())
+	case pcommon.ValueTypeDouble:
+		m.PutDouble(key, value.Double())
+	case pcommon.ValueTypeBool:
+		m.PutBool(key, value.Bool())
+	case pcommon.ValueTypeBytes:
+		m.PutEmptyBytes(key).FromRaw(value.Bytes().AsRaw())
+	case pcommon.ValueTypeMap:
+		value.Map().CopyTo(m.PutEmptyMap(key))
+	case pcommon.ValueTypeSlice:
+		value.Slice().CopyTo(m.PutEmptySlice(key))
+	case pcommon.ValueTypeEmpty:
+		m.PutEmpty(key)
+	default:
+		return fmt.Errorf("unsupported value type %s for key %q", value.Type().String(), key)
+	}
+	return nil
+}
+
+func extractPcommonValue(m pcommon.Map, path []string) (pcommon.Value, error) {
 	if len(path) == 0 {
-		return nil, errors.New("must provide at least one path item")
+		return pcommon.NewValueEmpty(), errors.New("must provide at least one path item")
 	}
-	obj, ok := v[path[0]]
+
+	val, ok := m.Get(path[0])
 	if !ok {
-		return nil, fmt.Errorf("provided object does not contain the path %v", path)
+		return pcommon.NewValueEmpty(), fmt.Errorf("provided object does not contain the path %v", path)
 	}
+
 	if len(path) == 1 {
-		return obj, nil
+		return val, nil
 	}
 
-	if o, ok := obj.(map[string]any); ok {
-		return extractValue(o, path[1:])
+	if val.Type() != pcommon.ValueTypeMap {
+		return pcommon.NewValueEmpty(), fmt.Errorf("provided object does not contain the path %v", path)
 	}
-	return nil, fmt.Errorf("provided object does not contain the path %v", path)
+
+	return extractPcommonValue(val.Map(), path[1:])
```

Comment:
my understanding is that deep recursion is unlikely here, so I kept it like this for readability...